Sprint 9 solution http api

DmitryFedoroff opened the pull request:
#comment Refactor task management system to include HTTP server and request handlers, create new adapters for LocalDateTime and Duration serialization, and enhance unit test coverage:

1. Introduce HttpTaskServer to handle HTTP requests for task management.
2. Define BaseHttpHandler as an abstract base class for all HTTP handlers.
3. Implement EpicsHandler, HistoryHandler, PrioritizedHandler, SubtasksHandler, and TasksHandler for handling respective HTTP requests.
4. Create NotFoundException for handling resource not found errors.
5. Implement LocalDateTimeAdapter for LocalDateTime serialization and deserialization.
6. Add DurationAdapter for Duration serialization and deserialization.
7. Refactor TaskManager to support new HTTP server and handlers.
8. Enhance SimpleTask class to support new features.
9. Improve TimeUtils methods for better date-time handling.
10. Develop unit tests for HttpTaskServer and its handlers.
11. Add unit tests for LocalDateTimeAdapter and DurationAdapter.
12. Update SimpleTaskTest to cover changes in SimpleTask.

Affected: HttpTaskServer, BaseHttpHandler, EpicsHandler, HistoryHandler, PrioritizedHandler, SubtasksHandler, TasksHandler, NotFoundException, LocalDateTimeAdapter, DurationAdapter, TaskManager, SimpleTask, TimeUtils, HttpTaskManagerTasksTest, HttpTaskManagerSubtasksTest, HttpTaskManagerPrioritizedTest, HttpTaskManagerHistoryTest, HttpTaskManagerEpicsTest, DurationAdapterTest, LocalDateTimeAdapterTest, SimpleTaskTest"

DmitryFedoroff committed:
refactor: Optimize task and history management for better performance and consistency

#comment Refactor various components of the task management application to enhance performance, ensure data consistency, and simplify code structure. Key changes include:
1. InMemoryTaskManager: Modify task deletion methods to remove tasks, subtasks, and epics from history when they are deleted. This ensures that the task history is accurate and up-to-date.
2. HistoryManager Interface: Introduce a new remove method to allow tasks to be explicitly removed from the history. This addition supports maintaining an accurate and concise history of task interactions.
3. InMemoryHistoryManager: Replace the LinkedList implementation with a custom doubly linked list to improve efficiency in managing task history. Implement methods such as linkLast, removeNode, and getTasks to handle history operations. Ensure that tasks are uniquely added to the history and properly managed when removed.
4. Main Class: Simplify the task creation and history verification process. Streamline task operations to focus on creating tasks, subtasks, and epics, followed by various history checks and deletions. This simplification makes the main application logic clearer and more maintainable.
5. Tests: Update existing tests and add new tests to cover the enhanced functionalities. Ensure comprehensive test coverage for task removal from history and verification of task deletions, validating the correctness and robustness of the implemented changes

Affected: InMemoryTaskManager, HistoryManager, InMemoryHistoryManager, Main, InMemoryHistoryManagerTest, InMemoryTaskManagerTest

DmitryFedoroff committed:
refactor: Improve code quality and test coverage

#comment Refactor code to improve clarity, remove redundant checks, and enhance test coverage:

1. Remove unnecessary viewCounts field from InMemoryHistoryManager.
2. Add null checks in the add method of InMemoryHistoryManager.
3. Remove duplicate node checks in the remove method of InMemoryHistoryManager.
4. Streamline InMemoryTaskManager by ensuring addToHistory handles null checks and implementing epic deletion from history.
5. Extend unit tests in InMemoryHistoryManagerTest and InMemoryTaskManagerTest to cover more scenarios, including removing tasks from different parts of history.

Affected: InMemoryHistoryManager, InMemoryTaskManager, InMemoryHistoryManagerTest, InMemoryTaskManagerTest

DmitryFedoroff committed:
refactor: Optimize task and history management

1. Add necessary empty lines between method definitions in interfaces to adhere to CheckStyle requirements and prevent style errors.
2. Update InMemoryHistoryManager to remove existing task nodes before adding new ones, ensuring unique task entries in history.
3. Modify InMemoryHistoryManager's getHistory method to use a separate method for task retrieval, improving code readability and maintainability.
4. Change methods in InMemoryTaskManager to add tasks to history upon retrieval instead of creation, ensuring accurate task tracking.
5. Optimize deleteAllEpics method in InMemoryTaskManager to ensure efficient removal of epics and associated subtasks, enhancing performance.

Affected: HistoryManager, TaskManager, InMemoryHistoryManager, InMemoryTaskManager

DmitryFedoroff committed:
refactor: Improve task management and history tracking

1. Enhance InMemoryTaskManager to include history updates on task addition and updates.
2. Simplify BaseTask equals and hashCode methods to only consider id field for equality and hashcode calculation.
3. Add validation in EpicTask to prevent adding itself as a subtask.
4. Add validation in Subtask class to prevent setting its id to its epicId.
5. Correct and enhance tests for InMemoryHistoryManager, improve test coverage for InMemoryTaskManager, update tests for BaseTask, and add validation tests for Subtask.

Affected: InMemoryTaskManager, BaseTask, EpicTask, Subtask, InMemoryHistoryManagerTest, InMemoryTaskManagerTest, BaseTaskTest, SubtaskTest

DmitryFedoroff committed:
Merge branch 'main' of github.com:DmitryFedoroff/java-kanban

DmitryFedoroff committed:
refactor: Enhance task manager with file-backed storage and improve test coverage

#comment Refactor task manager to include file-backed storage and add comprehensive tests:

1. Refactor InMemoryTaskManager to improve task management logic.
2. Introduce FileBackedTaskManager for persistent task storage using files.
3. Add ManagerSaveException to handle exceptions during file operations.
4. Create comprehensive tests for FileBackedTaskManager to ensure reliable functionality.

Affected: InMemoryTaskManager, FileBackedTaskManager, FileBackedTaskManagerTest, ManagerSaveException

DmitryFedoroff committed:
refactor: Restructure task management system for improved efficiency and reliability
#comment Refactor task management system to enhance performance and maintainability:

1. Optimize file I/O operations using BufferedWriter and BufferedReader
2. Standardize file format with consistent header and task structure
3. Implement custom toString() methods for each task type
4. Enhance taskFromString method to handle task types more robustly
5. Integrate history management into file operations
6. Strengthen error handling with specific exception usage
7. Expand test suite to cover edge cases and new functionality
8. Update task retrieval methods to ensure history persistence
9. Improve code structure and introduce constants for consistency

Affected: FileBackedTaskManager, SimpleTask, EpicTask, Subtask, InMemoryTaskManager, FileBackedTaskManagerTest

DmitryFedoroff committed:
Merge branch 'main' of github.com:DmitryFedoroff/java-kanban

DmitryFedoroff committed:
refactor: Improve task management and testing structure

#comment Refactor task management system and testing suite:

1. Introduce BaseTask, SimpleTask, EpicTask, and Subtask classes with properties such as ID, title, description, status, duration, and start time.
2. Implement InMemoryTaskManager and FileBackedTaskManager to handle tasks in memory and with file persistence respectively.
3. Develop EpicTask handling, including adding and removing subtasks, and automatic status updates based on subtask statuses.
4. Create HistoryManager and InMemoryHistoryManager to track recently viewed tasks.
5. Add task overlapping validation to ensure no two tasks overlap in time.
6. Extend task management features to include getting prioritized tasks by start time.
7. Write comprehensive unit tests for BaseTask, SimpleTask, EpicTask, Subtask, InMemoryTaskManager, FileBackedTaskManager, and InMemoryHistoryManager.
8. Refactor Main class to demonstrate task creation, updating, deletion, and viewing history.
9. Include ManagerSaveException for handling file persistence errors gracefully.

Affected: tasks.csv, Main, FileBackedTaskManager, InMemoryHistoryManager, InMemoryTaskManager, TaskManager, BaseTask, EpicTask, SimpleTask, Subtask, FileBackedTaskManagerTest, InMemoryHistoryManagerTest, InMemoryTaskManagerTest, TaskManagerTest, EpicTaskTest

DmitryFedoroff committed:
refactor: Improve code structure and efficiency across multiple classes

#comment Refactor codebase to enhance structure, readability, and performance:

1. Ensure the getPrioritizedTasks method is overridden in the InMemoryTaskManager class.
2. Optimize the isTaskOverlapping method by combining filter and anyMatch conditions to avoid double iteration.
3. Remove the getHistoryManager method from the TaskManager interface to maintain encapsulation and implementation details.
4. Add TimeUtils utility class for handling time formatting operations.
5. Utilize the TimeUtils.timeToString method in BaseTask for consistent time formatting and to eliminate code duplication.
6. Simplify the isOverlapping method in BaseTask by using lazy logical operators to enhance readability and performance.
7. Refactor TaskManagerTest to align with the changes in the TaskManager and InMemoryTaskManager classes.
8. Add TimeUtilsTest to verify the correctness of the time formatting utility methods.

Affected: TaskManager, InMemoryTaskManager, BaseTask, TimeUtils, TaskManagerTest, TimeUtilsTest

DmitryFedoroff committed:
Merge branch 'main' of github.com:DmitryFedoroff/java-kanban

DmitryFedoroff committed:
refactor: Update task management system with new HTTP server, handlers, adapters, and unit tests

#comment Refactor task management system to include HTTP server and request handlers, create new adapters for LocalDateTime and Duration serialization, and enhance unit test coverage:

1. Introduce HttpTaskServer to handle HTTP requests for task management.
2. Define BaseHttpHandler as an abstract base class for all HTTP handlers.
3. Implement EpicsHandler, HistoryHandler, PrioritizedHandler, SubtasksHandler, and TasksHandler for handling respective HTTP requests.
4. Create NotFoundException for handling resource not found errors.
5. Implement LocalDateTimeAdapter for LocalDateTime serialization and deserialization.
6. Add DurationAdapter for Duration serialization and deserialization.
7. Refactor TaskManager to support new HTTP server and handlers.
8. Enhance SimpleTask class to support new features.
9. Improve TimeUtils methods for better date-time handling.
10. Develop unit tests for HttpTaskServer and its handlers.
11. Add unit tests for LocalDateTimeAdapter and DurationAdapter.
12. Update SimpleTaskTest to cover changes in SimpleTask.

Affected: HttpTaskServer, BaseHttpHandler, EpicsHandler, HistoryHandler, PrioritizedHandler, SubtasksHandler, TasksHandler, NotFoundException, LocalDateTimeAdapter, DurationAdapter, TaskManager, SimpleTask, TimeUtils, HttpTaskManagerTasksTest, HttpTaskManagerSubtasksTest, HttpTaskManagerPrioritizedTest, HttpTaskManagerHistoryTest, HttpTaskManagerEpicsTest, DurationAdapterTest, LocalDateTimeAdapterTest, SimpleTaskTest

LexLippi commented on src/taskmanagement/http/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,129 @@
+            }
+        } else if (splitStrings.length == 4 && splitStrings[3].equals("subtasks")) {
+            Optional<Integer> epicId = getIdFromPath(exchange);
+            if (epicId.isPresent()) {
```
Для такого стоит использовать функционал `.ifPresentOrElse()`

DmitryFedoroff replied:
**Исправлено**: Устранена проверка `.isPresent`, заменена на использование `.ifPresentOrElse` для получения `epicId`.

LexLippi commented on src/taskmanagement/http/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,129 @@
+    private void handlePostEpics(HttpExchange exchange) throws IOException {
+        EpicTask epic = getEpicFromRequestBody(exchange);
+        try {
+            if (epic.getId() != 0) {
```
В таком случае лучше использовать не примитивный тип `int`, а объектный тип `Integer` и в таком случае сделать проверку на `null`

DmitryFedoroff replied:
**Исправлено**: Использован тип `Integer` вместо типа `int` для `ID` с проверкой на `null`.

LexLippi commented on src/taskmanagement/http/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    private void handleGetEpicById(HttpExchange exchange, Optional<Integer> epicId) throws IOException {
+        if (epicId.isPresent()) {
```
Аналогично стоит использовать `.ifPresentOrElse()`

DmitryFedoroff replied:
**Исправлено**: Устранена проверка `.isPresent()`, заменена на использование `.ifPresentOrElse()` для `handleGetEpicById()`.

LexLippi commented on src/taskmanagement/http/handlers/EpicsHandler.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    private void handleDeleteEpicById(HttpExchange exchange, Optional<Integer> epicId) throws IOException {
+        if (epicId.isPresent()) {
```
Про это уже писал

DmitryFedoroff replied:
**Исправлено**: Устранена проверка `.isPresent()`, заменена на использование `.ifPresentOrElse() для `handleDeleteEpicById()`.

LexLippi commented on src/taskmanagement/http/handlers/SubtasksHandler.java:
```diff
@@ -0,0 +1,65 @@
+    private void handlePostSubtasks(HttpExchange exchange) throws IOException {
+        Subtask subtask = getSubtaskFromRequestBody(exchange);
+        try {
+            if (subtask.getId() != 0) {
```
про это уже писал

DmitryFedoroff replied:
**Исправлено**: Использован тип `Integer` вместо типа `int` для `ID` с проверкой на `null`.

LexLippi commented on src/taskmanagement/http/handlers/TasksHandler.java:
```diff
@@ -0,0 +1,66 @@
+    private void handlePostTasks(HttpExchange exchange) throws IOException {
+        SimpleTask task = getTaskFromRequestBody(exchange);
+        try {
+            if (task.getId() != 0) {
```
Про это уже писал

DmitryFedoroff replied:
***Исправлено*: Использован тип `Integer` вместо типа `int` для `ID` с проверкой на `null`.

LexLippi reviewed (changes requested):
Привет!

Хорошая работа!

Есть пара небольших замечаний, которые необходимо исправить!

Желаю удачи!

DmitryFedoroff committed:
refactor: Improve code readability and adhere to best practices

#comment Refactor code to improve readability and adhere to best practices:

1. Refactor TimeUtils to use a ternary operator for null check in stringToTime method.
2. Update EpicsHandler to use ifPresentOrElse instead of isPresent for Optional handling in handleGetEpicById, handleDeleteEpicById, and handleGetEpicSubtasks methods.
3. Change EpicsHandler to use Integer instead of int for ID with null checks in handlePostEpics method.
4. Modify SubtasksHandler to use Integer instead of int for ID with null checks in handlePostSubtasks method.
5. Change TasksHandler to use Integer instead of int for ID with null checks in handlePostTasks method.
6. Ensure all related tests reflect these changes and run successfully.

Affected: TimeUtils, TimeUtilsTest, EpicsHandler, SubtasksHandler, TasksHandler, Subtask, BaseTask

DmitryFedoroff commented on src/taskmanagement/util/TimeUtils.java:
```diff
@@ -4,12 +4,19 @@
+        if (time.equals("null")) {
+            return null;
+        }
+        return LocalDateTime.parse(time, DATE_TIME_FORMATTER);
```
**Исправлено**: Использован тернарный оператор для проверки на `null` в методе stringToTime.

LexLippi commented:
Привет!

Хорошая работа!

Желаю удачи в следующих спринтах!
